### ISSUE TITLE
Elastic Sink Connector with UPSERT does not work when using Schemas with Timestamps

### ISSUE THREAD

**Issue by dallaswrege** [datamountaineer/stream-reactor#166]
Using the Elastic Sink Connector with UPSERT coniguration, there seem to be issues when specifying Schemas that include fields of type org.apache.kafka.connect.data.Timestamp.

I am using the following connector config:

```
    "config" :
      {
	"connector.class" : "com.datamountaineer.streamreactor.connect.elastic.ElasticSinkConnector",
	"tasks.max" : 1,
	"connect.elastic.url" : "localhost:9300",
	"connect.elastic.cluster.name" : "mycluster",
	"topics" : "basesubscriber",
	"connect.elastic.sink.kcql" : "UPSERT INTO basesubscriber SELECT subscriberID, customerType, msisdn, nationality, language, activationDay FROM basesubscriber PK subscriberID"
      }

```

The field "activationDay" is a Connect Data Timestamp field. Things work great if I omit that field from the KCQL. Including that timestamp field results in the errors shown below.

When data goes onto the topic, I see the following error from the sink connector:

```
[2017-04-25 09:12:35,660] ERROR Task elasticsearch_sink_basesubscriber-0 threw an uncaught and unrecoverable exception (org.apache.kafka.connect.runtime.WorkerSinkTask)
java.lang.ClassCastException: java.util.Date cannot be cast to java.lang.Long
	at org.apache.kafka.connect.data.Struct.getInt64(Struct.java:131)
	at com.datamountaineer.streamreactor.connect.schemas.StructFieldsExtractor.com$datamountaineer$streamreactor$connect$schemas$StructFieldsExtractor$$getFieldValue(StructFieldExtractor.scala:80)
	at com.datamountaineer.streamreactor.connect.schemas.StructFieldsExtractor$$anonfun$2.apply(StructFieldExtractor.scala:55)
	at com.datamountaineer.streamreactor.connect.schemas.StructFieldsExtractor$$anonfun$2.apply(StructFieldExtractor.scala:54)
	at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:241)
	at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:241)
	at scala.collection.Iterator$class.foreach(Iterator.scala:893)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
	at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
	at scala.collection.TraversableLike$class.flatMap(TraversableLike.scala:241)
	at scala.collection.AbstractTraversable.flatMap(Traversable.scala:104)
	at com.datamountaineer.streamreactor.connect.schemas.StructFieldsExtractor.get(StructFieldExtractor.scala:54)
	at com.datamountaineer.streamreactor.connect.elastic.ElasticJsonWriter$$anonfun$3$$anonfun$5.apply(ElasticJsonWriter.scala:94)
	at com.datamountaineer.streamreactor.connect.elastic.ElasticJsonWriter$$anonfun$3$$anonfun$5.apply(ElasticJsonWriter.scala:86)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.immutable.Set$Set1.foreach(Set.scala:94)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.AbstractSet.scala$collection$SetLike$$super$map(Set.scala:47)
	at scala.collection.SetLike$class.map(SetLike.scala:92)
	at scala.collection.AbstractSet.map(Set.scala:47)
	at com.datamountaineer.streamreactor.connect.elastic.ElasticJsonWriter$$anonfun$3.apply(ElasticJsonWriter.scala:86)
	at com.datamountaineer.streamreactor.connect.elastic.ElasticJsonWriter$$anonfun$3.apply(ElasticJsonWriter.scala:78)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.immutable.Map$Map1.foreach(Map.scala:116)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
	at com.datamountaineer.streamreactor.connect.elastic.ElasticJsonWriter.insert(ElasticJsonWriter.scala:78)
	at com.datamountaineer.streamreactor.connect.elastic.ElasticJsonWriter.write(ElasticJsonWriter.scala:68)
	at com.datamountaineer.streamreactor.connect.elastic.ElasticSinkTask$$anonfun$put$2.apply(ElasticSinkTask.scala:68)
	at com.datamountaineer.streamreactor.connect.elastic.ElasticSinkTask$$anonfun$put$2.apply(ElasticSinkTask.scala:68)
	at scala.Option.foreach(Option.scala:257)
	at com.datamountaineer.streamreactor.connect.elastic.ElasticSinkTask.put(ElasticSinkTask.scala:68)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:429)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.poll(WorkerSinkTask.java:250)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.iteration(WorkerSinkTask.java:179)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:148)
	at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:139)
	at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:182)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
[2017-04-25 09:12:35,661] ERROR Task is being killed and will not recover until manually restarted (org.apache.kafka.connect.runtime.WorkerSinkTask)

```
Full trace is attached here: [fullTrace.txt](https://github.com/datamountaineer/stream-reactor/files/955383/fullTrace.txt)

Note in the trace that I deleted a version of the elasticsearch_sink_basesubscriber connector at time 09:10:32 (a version that worked - which omitted the activationDay field) - and then I re-created the elasticsearch_sink_basesubscriber connector with the offending configuration at time 09:10:48.

Here is the schema in the schema registry:

```
home$ curl -X GET http://$KAFKA_HOST:8081/subjects/basesubscriber-value/versions/1; echo
{"subject":"basesubscriber-value","version":1,"id":26,"schema":"{\"type\":\"record\",\"name\":\"basesubscriber_carrierb\",\"fields\":[{\"name\":\"subscriberID\",\"type\":\"string\"},{\"name\":\"customerType\",\"type\":\"string\"},{\"name\":\"activationDay\",\"type\":{\"type\":\"long\",\"connect.version\":1,\"connect.name\":\"org.apache.kafka.connect.data.Timestamp\",\"logicalType\":\"timestamp-millis\"}},{\"name\":\"deactivationDay\",\"type\":[\"null\",{\"type\":\"long\",\"connect.version\":1,\"connect.name\":\"org.apache.kafka.connect.data.Timestamp\",\"logicalType\":\"timestamp-millis\"}],\"default\":null},{\"name\":\"msisdn\",\"type\":[\"null\",\"string\"],\"default\":null},{\"name\":\"device\",\"type\":[\"null\",\"string\"],\"default\":null},{\"name\":\"deviceDay\",\"type\":[\"null\",{\"type\":\"long\",\"connect.version\":1,\"connect.name\":\"org.apache.kafka.connect.data.Timestamp\",\"logicalType\":\"timestamp-millis\"}],\"default\":null},{\"name\":\"nationality\",\"type\":[\"null\",\"string\"],\"default\":null},{\"name\":\"language\",\"type\":[\"null\",\"string\"],\"default\":null},{\"name\":\"birthday\",\"type\":[\"null\",{\"type\":\"long\",\"connect.version\":1,\"connect.name\":\"org.apache.kafka.connect.data.Timestamp\",\"logicalType\":\"timestamp-millis\"}],\"default\":null},{\"name\":\"gender\",\"type\":[\"null\",\"string\"],\"default\":null},{\"name\":\"billGroup\",\"type\":[\"null\",\"int\"],\"default\":null},{\"name\":\"blacklist\",\"type\":[\"null\",\"boolean\"],\"default\":null},{\"name\":\"baseCountry\",\"type\":[\"null\",\"string\"],\"default\":null},{\"name\":\"baseLocation\",\"type\":[\"null\",\"string\"],\"default\":null},{\"name\":\"baseTimeZone\",\"type\":[\"null\",\"string\"],\"default\":null},{\"name\":\"valueLevel\",\"type\":[\"null\",\"string\"],\"default\":null},{\"name\":\"lastUpdateDate\",\"type\":{\"type\":\"long\",\"connect.version\":1,\"connect.name\":\"org.apache.kafka.connect.data.Timestamp\",\"logicalType\":\"timestamp-millis\"}},{\"name\":\"category\",\"type\":[\"null\",\"string\"],\"default\":null},{\"name\":\"valueSegment\",\"type\":\"int\"},{\"name\":\"marketingValues\",\"type\":\"int\"},{\"name\":\"subscriberStatus\",\"type\":[\"null\",\"string\"],\"default\":null},{\"name\":\"distributor\",\"type\":\"string\"}],\"connect.version\":257,\"connect.name\":\"basesubscriber_carrierb\"}"}

```

And here is data on the topic - as displayed with the Kafka Console Consumer:

```
evol@evol-VB:~/working/kafka/src/com/evolving/nglm/lib$ $CONFLUENT_KAFKA_HOME/bin/kafka-avro-console-consumer --bootstrap-server $KAFKA_HOST:9092 --property schema.registry.url=http://$KAFKA_HOST:8081 --property print.key=true --from-beginning --topic basesubscriber
"121255503352"	{"subscriberID":"121255503352","customerType":"Prepaid","activationDay":1451624400000,"deactivationDay":null,"msisdn":{"string":"121255503352"},"device":{"string":"iPhone6s"},"deviceDay":{"long":1475357400000},"nationality":{"string":"us"},"language":{"string":"en"},"birthday":null,"gender":{"string":"Female"},"billGroup":null,"blacklist":null,"baseCountry":{"string":"US"},"baseLocation":null,"baseTimeZone":{"string":"America/New_York"},"valueLevel":{"string":"Medium"},"lastUpdateDate":1493125922000,"category":{"string":"-"},"valueSegment":1,"marketingValues":1,"subscriberStatus":{"string":"Active"},"distributor":"Main"}

```

I am using the following versions:
- Confluent 3.2.0
- Stream Reactor 0.2.5 (kafka-connect-elastic-0.2.5-3.2.0-all.jar)

**Comment by dallaswrege** [comment]
Connect Worker configuration is attached here: [connect-001.txt](https://github.com/datamountaineer/stream-reactor/files/955441/connect-001.txt)

**Comment by andrewstevenson** [comment]
@dallaswrege Thanks for the report, we'll take a look. If you want to contribute a PR you're more than welcome!

**Comment by dallaswrege** [comment]
Looking into this further, we figured out that the problem is actually with StructFieldReactor.scala over in the kafka-connect-common repository. I had worked on fixing it, but I see that @stheppi has already committed changes to the master - and his fix is more thorough than mine would have been. :-)

You guys are on top of this. Nice work.

**Comment by stheppi** [comment]
@dallaswrege 
 I have submitted the fix on master is on 3.2. if you are using 3.1.2 it will be available tomorrow once i merge the changes from master.
